format/ComputeIR.fbs

```diff
@@ -0,0 +1,510 @@
+  limit: long;
+}
+
+// The kind of join being produced
+enum RelationalJoinType : int {
+  INNER,
+  LEFT,
+  RIGHT,
+  FULL,
+  SEMI,
+  ANTI,
+}
+
+/// A standard relational / SQL-style equijoin.
+///
+/// Providing no left/right columns produces the cross product of the two
+/// tables.
+table Join {
+  // TODO: complete and document
+  type: RelationalJoinType = INNER;
+
+  left: TableExpr (required);
+  right: TableExpr (required);
+
+  // The expression to use for joining `left` and `right` tables
+  on_expr: ArrayExpr; // a missing on_expr indicates a cross join.
+}
+
+/// A temporal join type
+///
+/// TODO: complete and document
+table AsOfJoin {
+  left: TableExpr (required);
+  right: TableExpr;
+
+  /// The column in the left expression to use for data ordering to determine
+  /// the "as of" time.
+  left_asof: ColumnReference (required);
+
+  /// The column in the right expression to use for data ordering to determine
+  /// the "as of" time. If the column name is the same as left_asof, may be
+  /// omitted.
+  right_asof: ColumnReference;
+
+  /// TODO: Define means of providing time deltas in this IR.
+  tolerance: Literal;
+
+  /// If true, the
+  allow_equal: bool = true;
+}
+
+/// An extension of as-of join which allows applying an aggregate function to
+/// the data falling within the indicated time interval.
+///
+/// TODO: Define semantics of "identity" window where all elements of window
+/// become a List<T> element in the result.
+table WindowJoin {
+  // TODO
+}
+
+// The order in which to sort rows.
+enum Ordering : uint8 {
+  ASCENDING,
+  DESCENDING,
+}
+
+// The way in which NULL values should be ordered when sorting.
+enum NullOrdering : uint8 {
+  FIRST,
+  LAST
+}
+
+/// An expression to use for sorting. The key expression determines the values
+/// to be used for ordering the table's rows.
+table SortKey {
+  key: ArrayExpr (required);
+  ordering: Ordering = ASCENDING;
+  null_ordering: NullOrdering;
+}
+
+/// A table-generating function.
+///
+/// TODO
+table TableFunction {
+  descr: FunctionDescr (required);
+
+  /// An optional output schema for the table function. If not provided, must
+  /// be determined by the IR consumer.
+  out_schema: Schema;
+}
+
+union TableOperation {
+  ExternalTable,
+  Project,
+  Filter,
+  Aggregate,
+  Limit,
+  Join,
+  TableFunction
+}
+
+/// An expression
+table TableExpr {
+  /// The operation that yields data.
+  op: TableOperation (required);
+
+  /// An optional explicit name for this table expression, to enable
+  /// unambiguous column references. If not set, the name can be inherited from
+  /// an antecedent table in some cases.
+  name: string;
+
+  /// Optional output schema. A schema can be serialized here for informational
+  /// purposes, or to provide a checkpoint/assertion to the IR consumer about
+  /// what you expect the schema to be at this point. Always requiring it would
+  /// increase the on-wire size of a Table
```

Comment:
I'm skeptical that we should be concerned about the on-wire size of `TableExpr` here as opposed to getting the semantics we want.